third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp

```diff
@@ -201,21 +202,368 @@ struct TritonRaiseBlockPointer
+            }
+            return WalkResult::advance();
           })
           .Case<triton::LoadOp, triton::StoreOp>([this](auto loadstore) {
-            if (failed(rewriteLoadStoreOp(loadstore)))
+            if (failed(rewriteLoadStoreOp(loadstore))) {
               loadstore->emitRemark(
                   "TritonRaiseToBlockPointer: Failed to rewrite");
-          });
+              return WalkResult::advance();
+            }
+            return WalkResult::skip();
+          })
+          .Case<scf::ForOp>([&](auto forOp) {
+            if (rewriteForOp(forOp).failed()) {
+              forOp->emitRemark(
+                  "TritonRaiseToBlockPointer: Failed to rewrite ForOp");
+              return WalkResult::advance();
+            }
+            return WalkResult::skip();
+          })
+          .Default([&](auto) { return WalkResult::advance(); });
+    });
+
+    return success();
+  }
+
+  LogicalResult rewriteForOp(scf::ForOp op) {
+    SmallVector<Value> newInitArgs;
+
+    SmallVector<std::pair<int, PtrState>, 5> initArgIndexState;
+    SmallVector<std::pair<int, PtrState>, 5> knownPtrsTmp;
+
+    llvm::SmallDenseMap<int, PtrState> initArgIndexMap;
+
+    OpBuilder builder(op);
+
+    // Create a new list of init args
+    for (auto [i, arg] : llvm::enumerate(op.getInitArgs())) {
+      auto mappedV = ptrMap.lookupOrNull(arg);
+      PtrState state;
+      if (mappedV) {
+        if (auto makeTensorPtrOp =
+                mappedV.getDefiningOp<triton::MakeTensorPtrOp>()) {
+          if (visitOperandMakeTensorPtr(makeTensorPtrOp, state, op.getLoc(),
+                                        builder, true)
+                  .succeeded()) {
+            newInitArgs.push_back(mappedV);
+            // Record the PtrState for later processing
+            initArgIndexState.push_back(std::make_pair(i, state));
+            continue;
+          }
+        } else if (auto addptrOp = mappedV.getDefiningOp<triton::AddPtrOp>()) {
+          // We always use tt.addptr for scalar pointers. If the defininig op is
+          // tt.addptr and we have a non-scalar pointer, something must have
+          // gone wrong with the pass.
+          assert(!isa<RankedTensorType>(addptrOp.getResult().getType()));
+          if (visitOperandAddptr(addptrOp, state, op.getLoc(), builder)
+                  .succeeded()) {
+            newInitArgs.push_back(mappedV);
+            // Record the PtrState for later processing
+            initArgIndexState.push_back(std::make_pair(i, state));
+            continue;
+          }
+        }
+      }
+      // If any of the analysis failed, or init arg is not pointer related or
+      // prior rewrite has failed. Pass as is
+      newInitArgs.push_back(arg);
+    }
+
+    // For each of the PtrState recorded in the last step, insert new
+    // instructions to describe offset and stride for each dimension and append
+    // them to init args
+    for (auto [i, state] : initArgIndexState) {
```

Comment:
```suggestion
    for (auto &[i, state] : initArgIndexState) {
```

Avoid copying `state`